fix(helm): add service, secret manifests in `init -o`

ryane opened the pull request:
When `helm init -o yaml|json` is run, the service and optional secret manifests
will be included in the output. The manifests are included in a Kubernetes
[List](https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#List) object.

This also includes some refactoring and consolidates how manifests are output in
`--dry-run` and removes some unnecessary marshaling between Kubernetes objects
and strings.

Fixes #3145

ryane committed:
fix(helm): add service, secret manifests in `init -o`

When `helm init -o yaml|json` is run, the service and optional secret manifests
will be included in the output. The manifests are included in a Kubernetes
[List](https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#List) object.

This also includes some refactoring and consolidates how manifests are output in
`--dry-run` and removes some unnecessary marshaling between Kubernetes objects
and strings.

Fixes #3145

prydonius commented:
Nice @ryane, although instead of the List you could output multiple JSON structures next to each other. https://github.com/kubernetes/kubernetes/blob/master/hack/testdata/multi-resource-json-modify.json

I'm concerned switching the YAML output to a List might break scripts relying on it (though that might not be that important).

ryane commented:
Yeah I wondered about that too. I didn't realize that a json object structure like that would be deployable but it does seem to work. I'm happy to update the pr if we think that is the better approach.

ryane committed:
don't wrap helm init -o/--dry-run output in List

Manifests in yaml are separated by `---`. In json, manifests are separated by
newlines (see
https://github.com/kubernetes/kubernetes/blob/master/hack/testdata/multi-resource-json.json)

ryane commented:
I added a commit to remove the wrapping of the resources in a List.

bacongobbler commented on cmd/helm/installer/install.go:
```diff
@@ -113,40 +114,65 @@ func createDeployment(client extensionsclient.DeploymentsGetter, opts *Options)
+	if err != nil {
+		return nil, err
+	}
+	dep.TypeMeta = metav1.TypeMeta{
```
shouldn't this be part of `generateDeployment`? :)

ryane replied:
Yeah I think it probably could be. If you look at the original code, the `TypeMeta` information is only included on the object when generating manifests for `-o` and `--debug` and is injected manually as strings: https://github.com/kubernetes/helm/blob/master/cmd/helm/init.go#L230. So, in order to minimize the potential impact, this PR follows that same pattern: TypeMeta is only included when generating the manifests as strings and not when actually sending the object to the Kubernetes api. 

That said, I don't see why including TypeMeta on the objects sent to the Kubernetes api would cause a problem. What do you think?

bacongobbler replied:
Ooh good point. I didn't fully see that `generateDeployment` was sorta/kinda double dipping: once for the actual object being shipped to kubernetes on `helm init`, and once for the string output. This is fine. :+1:

bacongobbler commented on cmd/helm/installer/install.go:
```diff
@@ -113,40 +114,65 @@ func createDeployment(client extensionsclient.DeploymentsGetter, opts *Options)
-// deployment gets the deployment object that installs Tiller.
-func deployment(opts *Options) (*v1beta1.Deployment, error) {
-	return generateDeployment(opts)
+// Deployment gets the deployment object that installs Tiller.
```
Based on our other comment thread, would it be fair to update this docstring to mention that this is only for string representations and should not be submitted to the API?

ryane replied:
yes, definitely. I'll commit shortly

ryane committed:
update docstrings on Deployment, Service, and Secret

ryane commented:
anything else needed here to get this included in an upcoming release? we'd love to have this for our next tiller upgrade. thanks!

bacongobbler commented:
nope! just needs another core maintainer to review this since it's a size/L

jstoja commented:
@jascott1 ? :)

jascott1 commented:
@jstoja First I apologize for taking so long to respond. I just built the PR and confirmed I can see your changes but when I run `helm init -o yaml` I do not see a secrets manifest. How do I see that?

jstoja commented:
@jascott1 I’m not the author of the PR, @ryane is.

But looking at the changes, you need to have TLS enabled to generate a secret manifest (see line 161 of https://github.com/kubernetes/helm/pull/3556/files/61e88cbc412a0a77e4eba3dc03b888cc18ea647e#diff-fe5fb50606e92e999fee7ed3b189e76e)

(Sorry I’m writing this from a smartphone)

jascott1 reviewed (approved):
Manually tested. LGTM

ryane commented:
thanks @jascott1 for getting this going again!
